cloudweatherreport/run.py

```diff
@@ -1,33 +1,231 @@
+        """
+        Run BundleTester and create or update the report accordingly.
+
+        :return: True if the test environment was provisioned and deployed
+        """
+        env_name = env.name
+        bundletester_output = StringIO()
+        self.args.output = bundletester_output
+        self.args.tests = test_plan.tests if test_plan else None
+        self.args.environment = env_name
+        self.args.reporter = 'json'
+        self.args.testdir = test_plan.bundle
+        if test_plan.bundle_file:
+            self.args.bundle = test_plan.bundle_file
+        status = tester.main(self.args)
+        result = model.SuiteResult.from_bundletester_output(
+            env.provider_name,
+            bundletester_output.getvalue())
+        result.bundle_yaml = get_bundle_yaml(status)
+        return result
+
+    def run_benchmarks(self, test_plan, env):
+        actions_client = get_versioned_juju_api().facades.Actions(env)
+        env_status = env.status()
+        benchmarks = []
+        for benchmark_plan in test_plan.benchmarks:
+            logging.info('Running benchmark {} on {} with params: {}'.format(
+                benchmark_plan.action,
+                benchmark_plan.unit,
+                benchmark_plan.params))
+            real_unit = find_unit(benchmark_plan.unit, env_status)
+            if not real_unit:
+                raise Exception("unit not found: {}".format(
+                    benchmark_plan.unit))
+            try:
+                result = run_action(
+                    actions_client,
+                    real_unit,
+                    benchmark_plan.action,
+                    action_param=benchmark_plan.params,
+                    timeout=3600)
+            except Exception as e:
+                logging.error('Action run failed: {}'.format(str(e)))
+                continue
+            composite = result.get('meta', {}).get('composite')
+            if not composite:
+                logging.error('Skipping benchmark missing composite key: '
+                              '{}'.format(benchmark_plan.action))
+                continue
+            composite.update({
+                'name': benchmark_plan.action,
+                'test_id': self.test_id,
+                'provider': env.provider_name,
+            })
+            benchmarks.append(model.Benchmark.from_action(composite))
+            logging.info('Benchmark completed.')
+        return benchmarks
 
 
-def entry():
+def entry_point():
     args = parse_args()
-    test_plans = read_file(args.test_plan, 'yaml')
-    test_plans = [test_plans] if isinstance(test_plans, dict) else test_plans
-    for test_plan in test_plans:
-        html_filename = main(args, test_plan)
-        print("Test result:\n  {}".format(html_filename))
-        file_open_with_app((html_filename))
+    processes = []
+    for controller in args.controllers:
+        processes.append(Runner(controller, args))
+        processes[-1].start()
```

Comment:
I'm not sure that I see this as useful.  If you're debugging a run, I can see running a single cloud, but I can't imaging wanting to run multiple clouds and not have them go in parallel.

Perhaps we could compromise by prefix the CWR log lines with the cloud that the log message applies to?